### DISCUSSION TITLE
Nacelle yaw controller

### DISCUSSION THREAD

**Discussion by MYMahfouz** [NatLabRockies/ROSCO#174] [General]
Hi,

The nacelle yaw controller currently does not account for the platform yaw in case of floating wind turbines.

In the older version with the [ROSCO_toolbox repo](https://github.com/NREL/ROSCO_toolbox). When the floating platform yawed, the nacelle was considered also yawed and the yaw controller was activated to force the nacelle to face the wind.

In the current version, the controller starts working only when the nacelle is yawed relative to the tower reference frame, and not relative to the fixed (inertial) reference frame.
For floating, when this is not considered this means there are cases where there will be constant yaw misalignment.

I was using the ROSCO_toolbox earlier, but I had problems understanding what the fast and slow low pass filters meant, and how to set them.

I hope my issue is clear, and thanks a lot for guiding me to this repo and answering my question earlier.

Best regards,
Youssef

**Comment by dzalkind** [comment]
> In the older version with the [ROSCO_toolbox repo](https://github.com/NREL/ROSCO_toolbox). When the floating platform yawed, the nacelle was considered also yawed and the yaw controller was activated to force the nacelle to face the wind.

I'm not sure this was the case.  That would result in a lot of yaw motion.  Could you show me where this is happening?

In the old version of the yaw controller, here are the default versions of the fast/slow cut-off frequencies: https://github.com/NREL/ROSCO/blob/ad15bcdd6d51e8b537630543d1b97b9048c19d7c/ROSCO_toolbox/utilities.py#L457

Here is the logic explaining the fast/slow yaw error: https://github.com/NREL/ROSCO/blob/ad15bcdd6d51e8b537630543d1b97b9048c19d7c/ROSCO/src/Controllers.f90#L248

**Comment by MYMahfouz** [comment]
I have results that I will show here, where I activated the yaw controller.  

I have made the error threshold very small in this simulation. I am aware that this is not realistic. But I also don't know what is the minimum threshold that I can set, I am searching about this in literature.

This simulation shown is a steady state simulation at 10m/s. The wind direction is fixed to 180 degrees. Without the yaw controller the steady state platform yaw is around 8 degrees.

I see that the nacelle is moving a lot and this movement will be much higher once a turbulent wind field is applied. I am not sure how often can the nacelle move and what are the limits, but I am also trying to find out at the moment form the literature. I haven't found anything about this problem for floating wind turbines. Did you see any paper discussing this?

I am using the yaw controller not to totally eliminate the platform yaw but just to decrease its effect. Do you have any other ideas or suggestions to do this? 

Thanks

![yaw_ctrl](https://user-images.githubusercontent.com/33868271/192593902-b5a742cf-b320-4ab1-a992-4242e8c6d1c0.PNG)

**Comment by dzalkind** [comment]
We think [this](https://iopscience.iop.org/article/10.1088/1742-6596/1037/3/032011) is a fairly standard yaw controller approach, which is why we implemented it here.  You could compare your controller's results to that.  In other projects, we have used yaw error (which translates to power loss), yaw actuator travel, and the number of yaw activations as measures.  Typically, we want to reduce all those measures, but there is usually a trade-off.  In most yaw controllers, no action is taken if the yaw error is less than 8 degrees because the power loss is not so great and the amount of wind direction variation is larger than that in turbulence.

For floating turbines, mooring lines provide yaw stiffness that reduces platform yaw motion.

**Comment by MYMahfouz** [comment]
Thanks a lot for the help. This is really useful.

Yes I agree that the mooring lines provide the yaw stiffness. However, the stiffness vary according to the wind direction and for some wind directions there might be higher yaw values than others, and I am trying to find a solution to decrease the yaw misalignment for these wind directions.

I want to decrease the yaw misalignment not only for the power losses, but also the yaw misalignment creates a crosswind force component on the rotor which also affects the overall floater displacement in the crosswind direction. So I want to decrease this displacement so I aim to decrease the yaw misalignment.

I will compare the controllers as you suggested. 
My last question: are there any plans to update the current version of the yaw controller currently implemented within ROSCO to account for the floater yaw? 


Thanks.

**Comment by dzalkind** [comment]
The short answer is yes, the co-dependence of varying yaw stiffness, mooring design, and turbine displacement are on our radar.  The timeline and scope of our ROSCO development efforts are still to be determined.  

However, I think the current controller does account for floater yaw because it still depends on the nacelle vane angle, which should include the platform yaw.  The new controller just includes a [deadband](https://github.com/NREL/ROSCO/blob/51629665d22f92770f861daeebcebd47ab32abb1/Test_Cases/IEA-15-240-RWT-UMaineSemi/DISCON-UMaineSemi.IN#L97) where it won't activate unless the vane angle is greater than some threshold for a certain period of time. 

I'm not sure there are any software issues here, so I'm going to convert this to a discussion to keep it open.

**Comment by MYMahfouz** [comment]
When i tried this version of ROSCO, I have set the deadband angle to 1 degree to check. 
The platform yaw reached 7 degrees and the yaw controller was not activated. 

It is my first time to go through the ROSCO controller source code, I have tried to follow which signal the controller reads to get the nacelle vane angle, but I couldn't figure it out. 
In case the signal uses the NacYaw or YawPos from OpenFAST as a nacelle vane angle, then this will not account for the floating platform yaw angle. 
As the NacYaw measures relative to the tower reference frame and not the inertial reference frame. 

Thanks for your reply.

**Comment by MYMahfouz** [comment]
Hi again,

Sorry for my confusion earlier. I have done some debugging and I believe I have found why the yaw controller doesn't get activated.

In these [line](https://github.com/NREL/ROSCO/blob/51629665d22f92770f861daeebcebd47ab32abb1/Test_Cases/IEA-15-240-RWT-UMaineSemi/DISCON-UMaineSemi.IN#L96) and [line](https://github.com/NREL/ROSCO/blob/51629665d22f92770f861daeebcebd47ab32abb1/Test_Cases/IEA-15-240-RWT-UMaineSemi/DISCON-UMaineSemi.IN#L97) in the DISCON.IN file, if the "Y_uSwitch" value is zero, the second value of the "Y_ErrThresh" is used and not the first one as mentioned in the discon and as shown in the code [here](https://github.com/NREL/ROSCO/blob/51629665d22f92770f861daeebcebd47ab32abb1/ROSCO/src/Controllers.f90#L309). 


I also saw that in the debugging file the output of the Nacelle vane and all other yaw angles related parameters is in degrees and not in rad. I think there is a typo in the units of the debugging file.

The Yaw error in the debugging file is always printing zero and not the correct value.

Thanks a lot for the help, I appreciate it a lot. 

Please keep me updated if any other changes for the controller regarding the yaw of the floater and the nacelle.

Best,
Youssef

**Reply by dzalkind** [reply]
Good catch on the yaw control docs.  I'm updating them here: #175

I'll look into the yaw error and see what I find.

**Reply by dzalkind** [reply]
Okay, I just pushed some more fixes to #175 and ran a case to verify it was working properly.  You'll need to set `Y_ControlMode` to 1.  Here is a case with an initial yaw offset.  NacVane and NacHeading are also good signals to look at.  To see them and other `LocalVars` in ROSCO, you'll need to set `LoggingLevel` to 2.

I hope this helps.  Please let me know if you find anything interesting!

![image](https://user-images.githubusercontent.com/65573423/193918823-ef5ee61a-316b-496c-8c04-90205dc20d09.png)

Here is a case in turbulence.  It looks a little more active that I'd expect, but that can be tuned.

![image](https://user-images.githubusercontent.com/65573423/193921491-83a56584-b2d0-47f4-b324-9da8708fadf3.png)

**Reply by MYMahfouz** [reply]
Thanks for doing the changes and the test. I didn't use this new update as I am working currently with just a .dll coupled to OpenFAST, I will try to compile this version and use it.

I am analyzing currently how to decrease the platform yaw effects with minimum nacelle yaw control events. I believe I will need the yaw error signal to be able to do such analysis so I will try to compile the new version.

Is there an already compiled .dll that I can use or this only comes with the next version of ROSCO? Is there a way to couple ROSCO to OpenFAST without needing a compiled/simulink version?

**Comment by MYMahfouz** [comment]
I compared the "NacHeadingtarget" to the "NacVane+Nacelle Heading" as shown in the image.

I have predicted that after filtering both signals will have the same amplitude or the "NacVane+Nacelle Heading" will be higher, but I am seeing the opposite. Is this what you would expect? 


Only platform yaw is my source of yaw misalignment, the wind direction doesn't change in my simulations.

![image](https://user-images.githubusercontent.com/33868271/194615270-9d212d54-8be9-4f46-b5ed-e2d69c4645ac.png)
